### PR TITLE
🆕: create module to create, update skill, infos and get players

### PR DESCRIPTION
**Pull request by MarcusSilvaDev**

### CODE REVIEW

**Comment by MarcusSilvaDev** [comment]
Here’s an integrated explanation of both the controller and its corresponding tests

---

### **PR: Addition of Player API Controller and Integration Tests**

This PR introduces the **PlayerController** and a comprehensive set of integration tests that validate the behavior of the **Player API** endpoints. Below is an overview of the controller functionality, followed by detailed test cases verifying its behavior.

#### **PlayerController Overview**

The **PlayerController** provides several endpoints for interacting with player data:

1. **Get All Players** (`GET api/Player`)
   - This endpoint retrieves all players from the system. It supports pagination to return a subset of player data.
   - **Test**: The `Get_All_Players` test validates that the API returns the correct number of players, matches the total count in the database, and ensures that the pagination works as expected.

2. **Get Player by Id** (`GET api/Player/{id}`)
   - This endpoint fetches a specific player by their unique ID.
   - **Test**: The `Get_Player_By_Id` test ensures that the data returned for a specific player matches the data in the database. The `Get_Player_By_Id_NotFound` test ensures that when a non-existent player ID is provided, the API returns a 404 status code.

3. **Create Player** (`POST api/Player`)
   - This endpoint allows for the creation of a new player with the provided details.
   - **Test**: The `Create_Player` test ensures that a new player can be successfully created via a `POST` request, and that the API responds with a 200 OK status code when the creation is successful.

4. **Update Player Skills** (`PATCH api/Player/{id}/skills`)
   - This endpoint updates the skills of an existing player, such as skill points for various abilities (e.g., sword, axe).
   - **Test**: The `Edit_Skills_Player` test verifies that the player's skill data can be updated successfully, returning a 200 OK status code.

5. **Update Player Information** (`PATCH api/Player/{id}/infos`)
   - This endpoint updates general player information, including stats like health, level, gender, and vocation.
   - **Test**: The `Edit_Infos_Player` test ensures that the player's information can be updated correctly, and verifies that the new values are persisted in the database.

#### **Test Details**

##### **Get Tests**

1. **Get All Players** (`Get_All_Players`):
   - **Arrange**: A player is created in the database to ensure there is data to retrieve. 
   - **Act**: A `GET` request is made to fetch all players, and the player created earlier is removed afterward to avoid interference with other tests.
   - **Assert**: The test checks that the response is not null, that the total number of records matches the database count, and that the pagination is correct.

2. **Get Players with Filter** (`Get_All_Players_with_Filter`):
   - **Arrange**: Two players are created for testing. The test ensures that the filtering works by requesting players by their name.
   - **Act**: A `GET` request is sent with a filter applied.
   - **Assert**: The test verifies that the filtered response contains the correct player and that the total number of players has decreased as expected.

3. **Get Player By Id** (`Get_Player_By_Id`):
   - **Arrange**: A player is created, and their ID is used to fetch their data via a `GET` request.
   - **Act**: A request is made to retrieve the player by ID.
   - **Assert**: The test validates that the returned data matches the player’s original attributes (e.g., name, account ID, skills).

4. **Get Player By Id when Not Found** (`Get_Player_By_Id_NotFound`):
   - **Arrange**: The test makes a `GET` request with a non-existent player ID.
   - **Act**: A request is made with an invalid ID.
   - **Assert**: The test checks that the response returns a 404 status code.

##### **Post Test**

1. **Create Player** (`Create_Player`):
   - **Arrange**: An account is created, and valid player creation data is prepared.
   - **Act**: A `POST` request is made to create the player.
   - **Assert**: The test checks if the creation request returns a 200 OK status code.

##### **Patch Tests**

1. **Edit Skills Player** (`Edit_Skills_Player`):
   - **Arrange**: A player is created, and new skill values are set for them.
   - **Act**: A `PATCH` request is sent to update the player's skills.
   - **Assert**: The test ensures that the response status code is 200 OK, indicating a successful update.

2. **Edit Infos Player** (`Edit_Infos_Player`):
   - **Arrange**: A player is created, and a comprehensive update request is prepared, including changes to their health, level, gender, and other stats.
   - **Act**: A `PATCH` request is made to update the player’s information.
   - **Assert**: The test checks the response status code and also verifies that the changes are correctly reflected in the database.

#### **Test Coverage and Validation**

- **Test Coverage**: This PR covers essential functionalities of the Player API, such as listing players, searching with filters, creating new players, and updating player details. It also includes error handling for scenarios where a player is not found.
- **Validation**: The tests validate both successful operations (e.g., creation and updates) and error scenarios (e.g., not found errors) to ensure that the PlayerController behaves correctly in different situations.
- **Next Steps**: As the API evolves, additional edge cases and validation tests can be added to ensure even more robust coverage.

---

This integrated explanation gives a comprehensive view of the **PlayerController** and its corresponding tests. It highlights the functionality of each endpoint and how the tests ensure that the API performs as expected under different conditions.

image

File: src/Database/NeoServer.Data/Repositories/Player/PlayerRepository.cs [hunk 1]
```diff
@@ -53,7 +54,7 @@ public async Task<List<PlayerOutfitAddonEntity>> GetOutfitAddons(int playerId)
     public async Task<PlayerEntity> GetPlayer(string playerName)
     {
         await using var context = NewDbContext;
-        return await context.Players.FirstOrDefaultAsync(x => x.Name.Equals(playerName));
+        return await context.Players.FirstOrDefaultAsync(x => x.Name.ToLower().Equals(playerName.ToLower()));
```
Comment (caioavidal) [inline src/Database/NeoServer.Data/Repositories/Player/PlayerRepository.cs]: please use invariant case to compare

File: src/WebAPI/Controllers/PlayerController.cs [hunk 2]
```diff
@@ -1,48 +1,57 @@
...
+        if (!response.IsSuccess) 
+            return UnprocessableEntity(response.ErrorMessage);
+        
+        return Ok(SuccessMessage.PlayerSkillsUpdated);
     }
-
-    [HttpPost]
-    public async Task<IActionResult> Post([FromBody] PlayerPostRequest player)
+    
+    [HttpPatch("{id}/infos")]
```
Comment (caioavidal) [inline src/WebAPI/Controllers/PlayerController.cs:47]: lets rename this endpoint to something more meaningful in future. for now it is good

**Review by caioavidal — APPROVED** [review]

**Review by caioavidal — CHANGES_REQUESTED** [review]
changes

File: src/WebAPI/Controllers/PlayerController.cs (hunk 2 above)
Comment (MarcusSilvaDev) [reply src/WebAPI/Controllers/PlayerController.cs:47]: This endpoint name was based the admin of tibia.
image

**Review by caioavidal — APPROVED** [review]
